Smooth transitions on parameter changes

elemoine opened the pull request:
This PR is an alternative to https://github.com/openlayers/ol3/pull/3660 created by @thomasmoelhave. The code includes code and suggestions from @thomasmoelhave.

The code is not ready for review yet. I need to review it myself and add unit tests.

elemoine commented:
Example visible here: http://erilem.net/ol3/dynamic-params/examples/smooth-params.html

bartvde commented:
nice to see you working on this again @elemoine 
If you need a review next week or so, ping me.

elemoine commented:
@thomasmoelhave, is the WMTS service used in the `smooth-params` example reliable? And what about the token key used to access the service?

thomasmoelhave commented:
@elemoine the server is reliable, it is one of our production tileservers. The token is perpetual and was made for this specific purpose. It can be used in the examples going forward, also publicly.

elemoine commented:
Thanks a lot for your answer.

elemoine commented:
I worked on the code again today, to, among other things, add tests. One problem I have is I know that my code doesn't work when the tile layer is reprojected (because its projection is not the same as the view projection). Making my code work in that case too would require more work, that I am not sure I can right now.

@bartvde, @thomasmoelhave, feel free to take a look at my code, and tell me what you think about the "does not work with reprojected layers" limitation.

bartvde commented on examples/smooth-params.html:
```diff
@@ -0,0 +1,23 @@
+    <input id="slider" type="range" value="10" step="0.25" max="100" min="-1" />
+    <span id="theinfo"></span>
+  </div>
+</di>
```
should be ```</div>```

elemoine replied:
Fixed.

bartvde commented on examples/smooth-params.js:
```diff
@@ -0,0 +1,126 @@
+  document.getElementById('theinfo').innerHTML = sliderVal + ' meters.';
+};
+
+updateSourceDimension(staticSource, 10);
```
why not simply setting this value in the source dimensions?

bartvde replied:
Ah I guess to have ```the info``` filled with the initial value?

elemoine replied:
Right.

bartvde commented on examples/smooth-params.js:
```diff
@@ -0,0 +1,126 @@
+  dimensions: {
+    'threshold': 100
+  },
+  dynamicDimensions: ['threshold']
```
I'm not sure I like the term dynamicDimensions. Dynamic versus static does not really convey what this does IMHO?

elemoine replied:
It may be clear already, but "dynamic dimensions" are dimensions that the user will change (with a time slider for example), and for which we will use interim tiles while new tiles are loading. But I agree with you when you say that the term doesn't really convey what this does. I just don't have a better name right now.

bartvde replied:
Yes this was clear, but I also couldn't really think of a better name, though I did consider ```smoothDimensions``` briefly.

bartvde replied:
Or maybe transitionDimensions?

bartvde replied:
Normally @tschaub is good with names, so ping @tschaub ;-)

elemoine replied:
We could also decide that dimensions are always "dynamic", and not introduce `dynamicDimensions` at all. But I decided against that when I initially worked on the patch, because I thought that there may be cases where displaying interim/old tiles is not at all acceptable.

bartvde replied:
Agreed, I also thought this might make for a sensible default.

elemoine replied:
Do you mean that parameters should be dynamic by default? Or do you agree with adding a `dynamicDimensions` option (or one with a better name)?

bartvde replied:
Parameters should be dynamic by default.

elemoine replied:
So in that case we'd need introduce a `staticDimensions` variable.

bartvde commented on src/ol/source/tileimagesource.js:
```diff
@@ -163,6 +163,36 @@ ol.source.TileImage.prototype.getTileCacheForProjection = function(projection) {
+ * @param {number} y Tile coordinate y.
+ * @param {number} pixelRatio Pixel ratio.
+ * @param {ol.proj.Projection} projection Projection.
+ * @param {string} key Key.
```
I think this parameter needs a bit more explaining, also why it needs to be set on the tile instance.

bartvde commented on src/ol/source/tilesource.js:
```diff
@@ -157,6 +157,14 @@ ol.source.Tile.prototype.getGutter = function() {
 
 
 /**
+ * @return {string} The dynamic parameters key.
```
This needs more docs IMHO.

bartvde commented on src/ol/structs/lrucache.js:
```diff
@@ -230,6 +230,18 @@ ol.structs.LRUCache.prototype.pop = function() {
+ol.structs.LRUCache.prototype.replace = function(key, value) {
+  var entry = this.entries_[key];
+  goog.asserts.assert(entry !== undefined, 'an entry exists for key %s', key);
+  this.get(key);
```
get has this same assertion. Can you explain why we need to call ```get``` here?

elemoine replied:
`get` updates the LRU structure (`oldest`, `newest`, etc.).

Maybe this would be better:

```js
ol.structs.LRUCache.prototype.replace = function(key, value) {
  this.get(key);  // update `newest_`
  this.entries_[key] = value;
};
```

elemoine replied:
I am also going to add a test for this new method.

bartvde commented on src/ol/tile.js:
```diff
@@ -44,6 +44,18 @@ ol.Tile = function(tileCoord, state) {
+  this.interimTile = null;
+
+  /**
+   * @type {string}
```
could do with a bit more docs

elemoine replied:
Yes, agree. Not easy to explain, but I'll try.

elemoine replied:
I've added more docs. I hope it clarifies a bit, but I am conscious it's not perfect.

bartvde commented:
@elemoine I don't think that not supporting tile reprojection needs to stand in the way of merging this PR. I've put a few small comments in the PR.

elemoine commented:
Thank you very much for your comments @bartvde! I didn't have time to look at them today, but I'll do it as soon as I can.

tschaub commented on src/ol/renderer/canvas/canvastilelayerrenderer.js:
```diff
@@ -317,18 +317,28 @@ ol.renderer.canvas.TileLayer.prototype.prepareFrame =
+        return tileState == ol.TileState.LOADED ||
+            tileState == ol.TileState.EMPTY ||
+            tileState == ol.TileState.ERROR && !useInterimTilesOnError;
+      });
```
Would it make sense to move this logic elsewhere?  For example, it might be nicer to be able to call `source.getTile()` and get the interim tile if the "main" one was not loaded (and yes, this would mean passing `useInterimTilesOnError` to `getTile`).

tschaub replied:
I see below where this "drawable" check is needed twice.

elemoine replied:
> For example, it might be nicer to be able to call source.getTile() and get the interim tile if the "main" one was not loaded (and yes, this would mean passing useInterimTilesOnError to getTile).

I thought about that. But `manageTilePyramid` is another [user](https://github.com/openlayers/ol3/blob/master/src/ol/renderer/layerrenderer.js#L320) of `getTile`.

tschaub commented on src/ol/renderer/canvas/canvastilelayerrenderer.js:
```diff
@@ -317,18 +317,28 @@ ol.renderer.canvas.TileLayer.prototype.prepareFrame =
+        tile = tile.interimTile;
+      }
+      goog.asserts.assert(tile);
+      if (drawableTile(tile)) {
```
Interesting that the DOM and WebGL separate the loaded/other case here.

elemoine replied:
I think it's related to the way the renderer draws "blank" tiles. I just reused the existing logic here.

tschaub commented:
Great to see this work picked up again @elemoine.  I hope to make time for a proper review.  Before merging, I think it would be nice to update the example so it doesn't highlight the bad (regular) case.

elemoine commented:
> Before merging, I think it would be nice to update the example so it doesn't highlight the bad (regular) case.

Indeed. I just changed the order. I am happy to try to make the example look nicer if there are suggestions. I have to admit that it's not the part I focused on.

elemoine commented:
@bartvde, I've taken your comments into account, but I am still not sure what to do with `dynamicDimensions` (or `staticDimensions`). I'll try to think more about it, but I am not sure I can come up with something better :)

marcjansen commented:
What about 'variableDimensions'?  Or 'changingDimensions'? Or 'movingDimensions'? Or 'flexibleDimensions'? 

From http://www.thesaurus.net/antonyms-for/Static & http://www.thesaurus.net/synonyms-for/Dynamic

marcjansen commented:
Volatile? Transient? Momentary? Impermanent? http://martinfowler.com/bliki/TwoHardThings.html

elemoine commented:
I gave this a bit more thinking. My new suggestion is to remove the `dynamicDimensions` option entirely, and make the new behavior (with interim tiles) non-configurable. If we want to make this configurable in the future we could add an `hardDimensions` option, with something like this for the documentation:

```js
/**
 * The list of dimensions that, when changed, invalidate the tiles in the cache. So when changing
 * one the "hard" dimensions the layer will disappear until new tiles are loaded.
 * @type {Array.<string>|undefined}
 */
olx.source.WMTSOptions.prototype.hardDimensions;
```

I'll modify my branch if you agree with making the new behaviour non-configurable.

bartvde commented:
I agree with making the new behaviour non configurable

elemoine commented:
Thanks for the quick response. I'll change my branch.

elemoine commented:
@bartvde, patch updated.

bartvde commented:
thanks @elemoine I'll have another look at it soon

bartvde commented on examples/smooth-params.html:
```diff
@@ -0,0 +1,18 @@
+---
+layout: example.html
+title: Smooth parameters example
```
maybe we should now rename this example to ```wmts-dimensions.html``` instead?

elemoine replied:
Yep, makes sense.

bartvde commented:
A put a small note on the example name, but LGTM @elemoine feel free to merge

elemoine commented:
I changed the name of the example. I also removed the `drawableTile` function and inlined the code in the DOM and WebGL renderers, where the `drawableTile` function was called only once. This is related to a comment @tschaub made.

thomasmoelhave commented:
It's great to see this idea get this much traction - thanks for the reworked patch @elemoine and for the reviews @bartvde and @tschaub! The new patch is significantly faster than your initial version.  If you are  interested in a little more thorough testing I will have time to look into updating our dev-version to this new version and test it out with our system next week. We have a large amount of layers with these types of parameters and if there is some hidden performance/feature regression from #3660 we are likely to find it.

elemoine commented:
Thanks @thomasmoelhave. More tests are always a good thing!

tschaub commented:
Nice work on this @elemoine.

thomasmoelhave commented:
Hi @elemoine et.al

Now that this has been merged I am not sure this pull request is the appropriate venue for this discussion, let me know if you want to move it elsewhere.

I tested  ol3 master and the dynamic WMTS dimension is significantly improved over the older unpatched OL3 version! However the performance is still significantly lower than our initial patch #3660 and this results in several rendering glitches. I adjusted our earlier example to use a window of the same size as the new dynamic WMTS dimension example to allow for an apples-to-apples comparison, so if you fire up `examples/wmts-dimensions.html` and compare it to http://scalgo.com/stuff/ol3/build/examples/smooth-params-big.html at the same window size you can compare them directly.

Zoom in on some area, e.g. the Netherlands and the eastern UK and drag the sea-level slider continuously, and at a decent speed, from one end of the slider range to the other. You should notice a  significant decrease in perceived performance. As a result, certain tiles "get stuck" for longer periods in the new version resulting in visible tile boundaries as shown in the attached figure. 
![output](https://cloud.githubusercontent.com/assets/179143/11244424/d3e5f280-8e0d-11e5-8f37-ed7b44d5a70a.png)
These glitches also appear in #3660, but they persist for much longer in this version. The combined result of this is a degradation of the user experience and the feeling that everything is slow. I have tested this in both Firefox and Chromium on Ubuntu 14.04.

At the moment I don't understand the structure of @elemoine's work to the extend that I can point to a particular thing that causes this issue, but it's probably worth looking into as the user experience could be better and this patch likely makes it much harder to apply the other patch at the same time. It might be something trivial, or something fundamental.

fredj commented:
> Now that this has been merged I am not sure this pull request is the appropriate venue for this discussion, let me know if you want to move it elsewhere.

You should open a new issue with the content of your comment. thanks
